kmod/src/server.c

```diff
@@ -2533,9 +2534,19 @@ static int server_commit_log_merge(struct super_block *sb,
 	if (ret < 0)
 		scoutfs_err(sb, "error %d committing log merge: %s", ret, err_str);
 
-	ret = scoutfs_server_apply_commit(sb, ret);
 	BUG_ON(ret < 0); /* inconsistent */
 
+	if (err < 0) {
```

Comment:
This all is pretty close, but still a bit off.   The big problem is that _apply_commit() is like an unlock.  We always have to call it after having called hold_commit or the commit_rwsem will still be held and server tasks will hang.  The other functional requirement is that we generally don't want to use BUG_ON while holding locks for the same reason, it kills the tasks and leaves locks held and the rest of the world hangs.  Then there's the stylistic preference of using err sparingly and when only necessary.   So putting all that together, I think we want: add a deleted bool initialized to false and set it to true after the initial deletion succeeds.   only use err to store the result of the apply_commit.   Leave the bug_on and apply where they are, but change the BUG conditino to ret < 0 && deleted.   Then after the bug set ret to err if ret was 0 so that it's sent as the net_response.   I think that's a minimal change that only triggers the bug if we're inconsistent.